x-pack/plugins/infra/public/utils/remote_state/remote_graphql_state.ts

```diff
@@ -0,0 +1,206 @@
+      ...state,
+      current: {
+        progress: 'idle',
+      },
+      last: {
+        result: 'failure',
+        reason: `${action.payload}`,
+        time: Date.now(),
+        parameters: {
+          operationKey,
+          variables: action.payload.params,
+        },
+      },
+    }))
+    .build();
+
+export const createGraphqlQueryEpic = <Data, Variables, Error = ApolloError>(
+  graphqlQuery: DocumentNode,
+  actionCreators: OperationActionCreators<Data, Variables, Error>
+): Epic<
+  ReduxAction,
+  ReduxAction,
+  any,
+  {
+    apolloClient$: Observable<InfraApolloClient>;
+  }
+> => (action$, state$, { apolloClient$ }) =>
+  action$.pipe(
+    filter(actionCreators.resolve.match),
+    withLatestFrom(apolloClient$),
+    switchMap(([{ payload: variables }, apolloClient]) =>
+      from(
+        apolloClient.query<Data>({
+          query: graphqlQuery,
+          variables,
+          fetchPolicy: 'no-cache',
+        })
+      ).pipe(
+        map(result => actionCreators.resolveDone({ params: variables, result })),
+        catchError(error => [actionCreators.resolveFailed({ params: variables, error })]),
+        startWith(actionCreators.resolveStarted(variables))
+      )
+    )
+  );
+
+export const createGraphqlStateSelectors = <State>(
+  selectState: (parentState: any) => GraphqlState<State> = parentState => parentState
+) => {
+  const selectData = createSelector(selectState, state => state.data);
+
+  const selectLoadingProgress = createSelector(selectState, state => state.current);
+  const selectLoadingProgressOperationInfo = createSelector(
+    selectLoadingProgress,
+    progress => (isRunningLoadingProgress(progress) ? progress.parameters : null)
+  );
+  const selectIsLoading = createSelector(selectLoadingProgress, isRunningLoadingProgress);
+  const selectIsIdle = createSelector(selectLoadingProgress, isIdleLoadingProgress);
+
+  const selectLoadingResult = createSelector(selectState, state => state.last);
+  const selectLoadingResultOperationInfo = createSelector(
+    selectLoadingResult,
+    result => (!isUninitializedLoadingResult(result) ? result.parameters : null)
+  );
+  const selectIsUninitialized = createSelector(selectLoadingResult, isUninitializedLoadingResult);
+  const selectIsSuccess = createSelector(selectLoadingResult, isSuccessLoadingResult);
+  const selectIsFailure = createSelector(selectLoadingResult, isFailureLoadingResult);
+
+  const selectLoadingState = createSelector(
+    selectLoadingProgress,
+    selectLoadingResult,
+    (loadingProgress, loadingResult) => ({
+      current: loadingProgress,
+      last: loadingResult,
+      policy: {
+        policy: 'manual',
+      } as LoadingPolicy,
+    })
+  );
+
+  return {
+    selectData,
+    selectIsFailure,
+    selectIsIdle,
+    selectIsLoading,
+    selectIsSuccess,
+    selectIsUninitialized,
+    selectLoadingProgress,
+    selectLoadingProgressOperationInfo,
+    selectLoadingResult,
+    selectLoadingState,
+  };
+};
```

Comment:
I like this way of generating all the redux thingies we will need for a GraphQL endpoint. :+1: No detailed comments right now, I might come back to that when I use it for the first time myself.